javacutil/src/main/java/org/checkerframework/javacutil/ElementUtils.java

```diff
@@ -239,6 +264,43 @@ public static boolean isCompileTimeConstant(Element elt) {
                 && ((VariableElement) elt).getConstantValue() != null;
     }
 
+    /**
+     * Checks whether a given element came from a source file.
+     *
+     * <p>By contrast, {@link ElementUtils#isElementFromByteCode(Element)} returns true if there is
+     * a classfile for the given element, even if there is also a source file.
+     *
+     * @param element the element to check, or null
+     * @return true if a source file containing the element is being compiled
+     */
+    public static boolean isElementFromSourceCode(@Nullable Element element) {
+        if (element == null) {
+            return false;
+        }
+        if (element instanceof ClassSymbol) {
+            return isElementFromSourceCodeImpl((ClassSymbol) element);
+        }
+        return isElementFromSourceCode(element.getEnclosingElement());
```

Comment:
This code uses `getEnclosingElement` (just like `isElementFromByteCode` immediately below it) because `enclosingClass` can return its input (that is, a class encloses itself), but `getEnclosingElement` always returns the actual enclosing element.